### DISCUSSION TITLE
Possible to generate arbitrary json objects?

### DISCUSSION THREAD

**Discussion by MTRNord** [loiclec/fuzzcheck-rs#33] [Q&A] [answered]
Hi I am using this in a rather unusual way currently to fuzz via HTTP.

I am sending json and one field can be a json object with basically any key variation.

So I am wondering what is the smartest way  to do this. Is a grammar based mutator the best way to do this?

**Comment by teymour-aldridge** [comment, marked as answer]
As this is a context-free grammar, I think so. The other way I know of is to use a recursive mutator to generate an enum (most likely with the same structure as [serde_json::Value](https://docs.rs/serde_json/latest/serde_json/enum.Value.html)) and then write a function to serialize it to a string (e.g. as https://github.com/bailion/compiler/blob/main/tests/generator/src/lib.rs does, but this becomes messy; JSON does have a much simpler grammar than that example).

There is a [chapter on grammar-based mutators](https://fuzzcheck.neocities.org/tutorial2.html) in the guide.

**Reply by MTRNord** [reply]
Thanks I will give it a try with a grammar then. I tried to use make_mutator on the serde_json::Value but as thats a foreign type it makes rust rather unhappy:

```
error[E0117]: only traits defined in the current crate can be implemented for types defined outside of the crate
  --> src/types.rs:59:1
   |
59 | / fuzzcheck::make_mutator! {
60 | |     name: ValueMutator,
61 | |     recursive: true,
62 | |     default: true,
63 | |     type:
64 | |         pub enum Value {
   | |                  ----^
   | |__________________|___|
   |                    |   impl doesn't use only types from inside the current crate
   |                    `Value` is not defined in the current crate
   |
   = note: define and implement a trait or new type instead
   = note: this error originates in the macro `fuzzcheck::make_mutator` (in Nightly builds, run with -Z macro-backtrace for more info)

For more information about this error, try `rustc --explain E0117`.
error: could not compile `matrix-fuzz` due to previous error
```

**Reply by teymour-aldridge** [reply]
Yes - the easiest way here (well, the second-easiest after using an [AST mutator](https://docs.rs/fuzzcheck/latest/fuzzcheck/mutators/grammar/struct.ASTMutator.html) and the [grammar module](https://docs.rs/fuzzcheck/latest/fuzzcheck/mutators/grammar/index.html)) would be to create a type with the same structure as `serde_json::Value` which could then be transformed into `serde_json::Value` recursively.

**Reply by MTRNord** [reply]
Hm turns out I cant use grammar here as I use structs for stuff :D This is rather annoying but oh well it could be worse. Guess I go the way of converting it.

**Reply by MTRNord** [reply]
Oh and yes I tried the grammar anyway. I then got stuck at converting it to actual json instead of a string escaped json. As serde_json (rightfully) does not allow control characters in the json. While this is nice for actual usage I want to have them encoded :) Even if this is invalid json.

**Reply by MTRNord** [reply]
I now went with a rather hacky approach of having in the type that fuzzcheck generates 2 vectors generating strings which I'm then using a From impl convert to a type having a hashmap. And as long as the sizes are lining up I should be able to generate the hashmap. If they dont line up I resize it as required.

This is not nice but it at least gets the job done for now.

**Reply by loiclec** [reply]
hi @MTRNord 👋 I am working on a solution for this problem this weekend. I will post updates here when I have something to share :)

**Reply by teymour-aldridge** [reply]
> Oh and yes I tried the grammar anyway. I then got stuck at converting it to actual json instead of a string escaped json. As serde_json (rightfully) does not allow control characters in the json. While this is nice for actual usage I want to have them encoded :) Even if this is invalid json.

@MTRNord I'm not 100% sure about what you mean here - would I be correct in interpreting this to mean that although the characters `"` and `\` should be escaped, you want to generate (invalid) JSON where they are not escaped?

I wrote a small grammar which generates only valid (but not the full set of) arbitrary JSON [here](https://git.sr.ht/~teymour/fuzzcheck_generators/) - the [code is here](https://git.sr.ht/~teymour/fuzzcheck_generators/tree/). I also wrote an implementation of the approach I described (in the same repository) to generate serde_json::Value items.

**Reply by teymour-aldridge** [reply]
> I now went with a rather hacky approach of having in the type that fuzzcheck generates 2 vectors generating strings which I'm then using a From impl convert to a type having a hashmap. And as long as the sizes are lining up I should be able to generate the hashmap. If they dont line up I resize it as required.

> This is not nice but it at least gets the job done for now.

Apart from the solution which @_loiclec is working on, one temporary workaround to do this would be to generate a `Vec<(K, V)>` where `K` and `V` stand for the key and the value types respectively and then use `collect` to build a HashMap/IndexMap etc.

**Reply by MTRNord** [reply]
> > I now went with a rather hacky approach of having in the type that fuzzcheck generates 2 vectors generating strings which I'm then using a From impl convert to a type having a hashmap. And as long as the sizes are lining up I should be able to generate the hashmap. If they dont line up I resize it as required.
> 
> > This is not nice but it at least gets the job done for now.
> 
> Apart from the solution which @_loiclec is working on, one temporary workaround to do this would be to generate a `Vec<(K, V)>` where `K` and `V` stand for the key and the value types respectively and then use `collect` to build a HashMap/IndexMap etc.

Oh. Yeah that's much smarter 👀😅😬 oops. Will change that 😄thanks!

**Reply by MTRNord** [reply]
> > Oh and yes I tried the grammar anyway. I then got stuck at converting it to actual json instead of a string escaped json. As serde_json (rightfully) does not allow control characters in the json. While this is nice for actual usage I want to have them encoded :) Even if this is invalid json.
> 
> @MTRNord I'm not 100% sure about what you mean here - would I be correct in interpreting this to mean that although the characters `"` and `\` should be escaped, you want to generate (invalid) JSON where they are not escaped?
> 
> I wrote a small grammar which generates only valid (but not the full set of) arbitrary JSON [here](https://git.sr.ht/~teymour/fuzzcheck_generators/) - the [code is here](https://git.sr.ht/~teymour/fuzzcheck_generators/tree/). I also wrote an implementation of the approach I described (in the same repository) to generating serde_json, but unfortunately I think I've uncovered another bug in Fuzzcheck (#34) which prevents this from working. 

While correct json is for most cases fine it is interesting to see what happens if the program gets invalid data. So yeah invalid json is interesting to some extent. Because while I expect the program to handle json correctly I would like to make sure that the fuzzer at least considers breaking it using invalid json. Just for that little more coverage of possibilities.

**Reply by MTRNord** [reply]
To add some more context: I am talking to an external http endpoint here and throw json against it. While fuzzcheck isn't meant to work for that it does yield usually better results than afl.rs do. So while it is not perfect it is more efficient :)

**Reply by teymour-aldridge** [reply]
> While correct json is for most cases fine it is interesting to see what happens if the program gets invalid data. So yeah invalid json is interesting to some extent. Because while I expect the program to handle json correctly I would like to make sure that the fuzzer at least considers breaking it using invalid json. Just for that little more coverage of possibilities.

You might find some of the approaches taken in [this article](https://comby.dev/blog/2022/04/11/comby-decomposer-compiler-fuzzing) interesting for generating kind-of-valid structured text.

> To add some more context: I am talking to an external http endpoint here and throw json against it. While fuzzcheck isn't meant to work for that it does yield usually better results than afl.rs do. So while it is not perfect it is more efficient :)

That's interesting; is this web service written in Rust (could Fuzzcheck obtain coverage data for it)?

**Reply by MTRNord** [reply]
> > While correct json is for most cases fine it is interesting to see what happens if the program gets invalid data. So yeah invalid json is interesting to some extent. Because while I expect the program to handle json correctly I would like to make sure that the fuzzer at least considers breaking it using invalid json. Just for that little more coverage of possibilities.
> 
> You might find some of the approaches taken in [this article](https://comby.dev/blog/2022/04/11/comby-decomposer-compiler-fuzzing) interesting for generating kind-of-valid structured text.

Thanks I will give it a read 👍

> 
> > To add some more context: I am talking to an external http endpoint here and throw json against it. While fuzzcheck isn't meant to work for that it does yield usually better results than afl.rs do. So while it is not perfect it is more efficient :)
> 
> That's interesting; is this web service written in Rust (could Fuzzcheck obtain coverage data for it)?

The webservice is https://github.com/matrix-org/synapse :) so sadly no rust. (and my python and fuzzing knowledge is not advanced enough to write fuzzer in python :))

Fuzzcheck still found some minor missing validation things and a potential security issue. Which is why "somewhat invalid" data seems very interesting

**Reply by MTRNord** [reply]
Oh and to add on the webservice part: I plan to probably expand this to other implementations of the matrix protocol :) One of them is even rust but as it is fully async it as well is kinda annoying to fuzz :D

**Reply by teymour-aldridge** [reply]
> Oh and to add on the webservice part: I plan to probably expand this to other implementations of the matrix protocol :) One of them is even rust but as it is fully async it as well is kinda annoying to fuzz :D

I imagine that it would be possible for fuzzcheck to have access to the coverage data for that program, however, which would improve performance markedly (I would guess).

p.s. I also fixed the mutator for `serde_json::Value` (at https://git.sr.ht/~teymour/fuzzcheck_generators, also now at [this Github mirror](https://github.com/teymour-aldridge/fuzzcheck_generators))

**Reply by MTRNord** [reply]
Thanks for writing that! I will try it as soon as it is a lower temp in my room :) I am melting currently :D

**Reply by MTRNord** [reply]
Thanks for writing it :) It works great!

Even when using it with `#[field_mutator(ValueMutator = {json_value_mutator()})]` :)

**Reply by loiclec** [reply]
Wow, thanks @teymour-aldridge for creating the mutator! :-) It's pretty much what I would have written as well 😄 My only comment is that maybe a `FilterMutator` could have been used to rule out strings with control characters instead of a `MapMutator`. It would probably be faster and the complexity would not have to be recomputed manually. I am only mentioning this in case that interests you, but I don't think you have to change anything since your solution works :) Note also that creating an `InternalJsonValue` type is not strictly necessary, since the `make_mutator!` macro can operate on foreign types as well.

On my side, I saw a few things I could write or improve in fuzzcheck to make generating Json values better:

1. A `f32`/`f64` mutator (this is 90% done on my computer)
2. A `HashMap`/`BTreeMap`/`serde_json::Map` mutator that isn't built on top of a `Vec<(K, V)>` mutator (I have started working on this as well. But note that nowadays I can only work on fuzzcheck on evenings/weekends, so I don't know when it will be ready).
3. The ability to add generic parameters / where clauses to the `make_mutator!` macro (not started yet). This would allow one to derive a mutator such as `JsonMutator<KeyMutator, StringMutator, etc.>`, which I think would be useful.

I am quite surprised to hear that fuzzcheck works for your use case @MTRNord . Do you use any kind of feedback (i.e. a `Sensor` and `Pool`) to direct the fuzzer towards interesting Json values?

Finally, I am curious about how both of you use fuzzcheck, and what I can do to make things easier. So if you'd like to have a call/chat about it (or anything else), feel free to email me at loiclecrenier@icloud.com and we can plan it :-)

**Reply by MTRNord** [reply]
> I am quite surprised to hear that fuzzcheck works for your use case @MTRNord . Do you use any kind of feedback (i.e. a `Sensor` and `Pool`) to direct the fuzzer towards interesting Json values?

Currently, not using that, but I am in the code itself disabling the false positives and not serializing values that are false positive cases. (As for example, things where valid validation of the input already happens). I probably should look into feedback instead 👍 

The code is at https://github.com/MTRNord/matrix-fuzz/blob/main/src/lib.rs#L118 It is a little awkward currently as I do force certain codepaths to favor results that are more likely to get accepted by the server. So I discard some of the fuzzing content and replace it.

I guess it does somewhat skew the results, though.  But it yielded some results already, while the afl target (disabled as arbitrary cannot do arbitrary json_server::Value either :D) had no results in the same amount of time and was noticeably slower than fuzzcheck. Afl.rs or arbitrary also seem to not generate enough data, as usually the json ended up just having mostly empty fields.

**Reply by teymour-aldridge** [reply]
I don't know if the current sensors are hardcoded into Fuzzcheck, but if not you might be able to write a custom one using `coverage.py` (or some other similar tool) to gather coverage data from a Python codebase.

Also, I have now [published `fuzzcheck_serde_json_generator` on crates.io](https://docs.rs/fuzzcheck_serde_json_generator/latest/fuzzcheck_serde_json_generator/), and updated it to generate escape characters (I forgot that `serde_json` automatically escapes escape characters when it serializes JSON).

**Reply by MTRNord** [reply]
> I don't know if the current sensors are hardcoded into Fuzzcheck, but if not you might be able to write a custom one using `coverage.py` (or some other similar tool) to gather coverage data from a Python codebase.
> 

hm interesting... Need to dive into this as I cannot imagine yet how I get from coverage in another language to getting feedback for the last HTTP request I made :D It seems like dark magic to me currently ^^

**Reply by MTRNord** [reply]
Just checked. coverage.py is already what they use for the e2e tests. So thats good news. :) So it will work. Just the magic of hooking it all together will be the step thats likely more complex

**Reply by teymour-aldridge** [reply]
> Just checked. coverage.py is already what they use for the e2e tests. So thats good news. :) So it will work. Just the magic of hooking it all together will be the step thats likely more complex

I have a scheme I _think_ will work, but won't have time to try and implement it before the weekend.

**Reply by loiclec** [reply]
It is possible indeed to create a custom `Sensor`/`Pool` combo that guides the fuzzer using code coverage collected from somewhere else. Although I'm a bit worried about the performance of it. As always, don't hesitate to contact me for help :-) 
Note also that it doesn't have to be guided by code coverage, it could be some other kind of feedback as well.

**Reply by teymour-aldridge** [reply]
My plan was to use Python's introspection to rewrite the synapse programme at runtime to provide coverage data to Fuzzcheck after every request, and then have a Fuzzcheck sensor and pool to interpret this data.

edit: I think the easiest method is to spin up a new CPython instance for each fuzz test, to start recording the coverage data at the start, and then at the end to extract the coverage data and send it to Fuzzcheck.

**Reply by MTRNord** [reply]
> Although I'm a bit worried about the performance of it.

I am kinda at the point where it doesnt really matter anyway as I can only do like 15Hz of requests anyway to the server as I am limited by the speed of the process. (It is running in a single process mode. I could add workers to it, but it just complicates stuff a lot). The things I am hitting are already at a complexity where the computation of the request tends to need 300-800ms depending on what exactly is sent. (Matrix is a distributed/federating chat protocol, so it has to do some rather complex calculations for that. Though arguably and based on data, most of this is just python being slow. There are go and rust implementations which are much faster but compared not feature complete)

**Reply by teymour-aldridge** [reply]
You might find the `--max-complexity <a number>` flag helpful, amongst the other suggestions [here](https://github.com/loiclec/fuzzcheck-rs/issues/31#issuecomment-1200470065) - lower down in the issue, not the part about the now fixed bug
